### PR TITLE
api: add pagination support

### LINKED ISSUE #246: Add pagination support
The feature will be available in the next Tarantool release. See the implementation:

**https://github.com/tarantool/tarantool/issues/7637**
https://github.com/tarantool/tarantool/issues/7639
https://github.com/tarantool/tarantool/pull/7763

### PR DESCRIPTION
**Pull request by oleg-jukovec**
A user could fetch a position of a last tuple using a new method of the SelectRequest type:

```Go
selectRequest = selectRequest.FetchPos(true)
```

The position will be stored in a new field of the Response type:

```Go
Response.Pos
```

A user could specify a tuple from which selection must continue or its position with a new method of the SelectRequest type:

```Go
selectRequest = selectRequest.After([]interface{}{23})
```

or

```Go
selectRequest = selectRequest.After(resp.Pos)
```

In action it looks like:

```Go
req := NewSelectRequest(space).Key(key).Limit(10).FetchPos(true)
for condition {
    resp, err := conn.Do(req).Get()
    // ...
    req = req.After(resp.Pos)
}
```

1. https://github.com/tarantool/tarantool/issues/7639
I didn't forget about (remove if it is not applicable):

- [x] Tests (see [documentation](https://pkg.go.dev/testing) for a testing package)
- [x] Changelog (see [documentation](https://keepachangelog.com/en/1.0.0/) for changelog format)
- [x] Documentation (see [documentation](https://go.dev/blog/godoc) for documentation style guide)

Related issues:

Closes #246

### CODE REVIEW

File: .github/workflows/testing.yml [hunk 1]
```diff
@@ -50,17 +50,43 @@ jobs:
...
+          path: "/opt/tarantool"
+          key: cache-latest-${{ env.LATEST_COMMIT }}
+
+      - name: Setup Tarantool 2.x latest
+        if: matrix.tarantool == '2.x-latest' && steps.cache-latest.outputs.cache-hit != 'true'
+        run: |
+          # mkdir could be removed after:
+          # https://github.com/tarantool/tt/issues/282
+          sudo mkdir -p /opt/tarantool
+          sudo tt install tarantool=master
```
Comment (DifferentialOrange) [inline .github/workflows/testing.yml:85]: It is theoretically possible that Tarantool master was updated between the steps. Maybe it's better to install `LATEST_COMMIT` here.

File: .github/workflows/testing.yml [hunk 2]
```diff
@@ -50,17 +50,43 @@ jobs:
...
+
+      - name: Add Tarantool 2.x latest to PATH
```
Comment (DifferentialOrange) [inline .github/workflows/testing.yml:87]: `tt install tarantool` doesn't set up path?

File: request.go [hunk 3]
```diff
@@ -18,13 +18,13 @@ func fillSearch(enc *encoder, spaceNo, indexNo uint32, key interface{}) error {
...
 
-func fillIterator(enc *encoder, offset, limit, iterator uint32) {
+func fillIterator(enc *encoder, offset, limit, iterator uint32) error {
 	encodeUint(enc, KeyIterator)
 	encodeUint(enc, uint64(iterator))
 	encodeUint(enc, KeyOffset)
 	encodeUint(enc, uint64(offset))
 	encodeUint(enc, KeyLimit)
-	encodeUint(enc, uint64(limit))
+	return encodeUint(enc, uint64(limit))
```
Comment (DifferentialOrange) [inline request.go:27]: Why it's the only field validated in `fillIteratior`?

File: request.go [hunk 4]
```diff
@@ -35,10 +35,50 @@ func fillInsert(enc *encoder, spaceNo uint32, tuple interface{}) error {
...
+		if err := enc.EncodeBool(fetchPos); err != nil {
+			return err
+		}
+	}
+	if after != nil {
+		if pos, ok := after.([]byte); ok {
+			if err := encodeUint(enc, KeyAfterPos); err != nil {
+				return err
+			}
+			if err := enc.EncodeString(string(pos)); err != nil {
```
Comment (DifferentialOrange) [inline request.go:69]: What's vmihailenco/msgpack rules for MP_BIN?

File: request.go [hunk 5]
```diff
@@ -716,14 +758,37 @@ func (req *SelectRequest) Key(key interface{}) *SelectRequest {
 	return req
 }
 
+// FetchPos determines whether to fetch positions of tuples. A position
```
Comment (DifferentialOrange) [inline request.go:761]: Isn't it more like "a position of the last tuple"?

File: request.go [hunk 6]
```diff
@@ -716,14 +758,37 @@ func (req *SelectRequest) Key(key interface{}) *SelectRequest {
...
+// descriptor will be saved in Response.Pos value.
+//
+// Requires Tarantool >= 2.11.
+//
+// Note: default value is false.
```
Comment (DifferentialOrange) [inline request.go:766]: Shouldn't we also have "supported since 1.11.0" here?

File: request.go [hunk 7]
```diff
@@ -716,14 +758,37 @@ func (req *SelectRequest) Key(key interface{}) *SelectRequest {
...
+	req.fetchPos = fetch
+	return req
+}
+
+// After must contain a tuple from which selection must continue or its
+// position (a value from Response.Pos).
+//
+// Requires Tarantool >= 2.11.
+//
+// Note: default value in nil.
```
Comment (DifferentialOrange) [inline request.go:777]: And here

File: response.go [hunk 8]
```diff
@@ -7,9 +7,12 @@ import (
 type Response struct {
 	RequestId uint32
 	Code      uint32
-	Error     string // error message
-	// Data contains deserialized data for untyped requests
-	Data     []interface{}
+	// Error message.
```
Comment (DifferentialOrange) [inline response.go:10]:
```suggestion
	// Error contains error message.
```
?

File: tarantool_test.go [hunk 9]
```diff
@@ -2397,6 +2447,112 @@ func TestClientRequestObjects(t *testing.T) {
 	}
 }
 
+func testConnectionDoSelectRequestPrepare(t *testing.T, conn Connector) {
+	t.Helper()
+
+	for i := 1010; i < 1020; i++ {
+		deleteReq := NewDeleteRequest(spaceName).Key([]interface{}{uint(i)})
+		insertReq := NewInsertRequest(spaceName).Tuple(
```
Comment (DifferentialOrange) [inline tarantool_test.go:2455]: Why not use a single replace request?

File: tarantool_test.go [hunk 10]
```diff
@@ -2397,6 +2447,112 @@ func TestClientRequestObjects(t *testing.T) {
...
+		t.Fatalf("Failed to Select: %s", err.Error())
+	}
+	if resp == nil {
+		t.Fatalf("Response is nil after Select")
+	}
+	if resp.Pos != nil {
+		t.Errorf("Response should not have a position")
+	}
+	if len(resp.Data) != 10 {
+		t.Fatalf("Response Data len %d != 9", len(resp.Data))
```
Comment (DifferentialOrange) [inline tarantool_test.go:2487]:
```suggestion
		t.Fatalf("Response Data len %d != 10", len(resp.Data))
```

File: tarantool_test.go [hunk 11]
```diff
@@ -2397,6 +2447,112 @@ func TestClientRequestObjects(t *testing.T) {
...
+		t.Fatalf("Failed to Select: %s", err.Error())
+	}
+	if resp.Pos == nil {
+		t.Fatalf("A response must have a position descriptor")
+	}
+	if len(resp.Data) != 2 {
+		t.Fatalf("Response Data len %d != 2", len(resp.Data))
+	}
+
+	for i := 0; i < 2; i++ {
```
Comment (DifferentialOrange) [inline tarantool_test.go:2530]: To be honest, this code is confusing. If you want to evade copypaste, maybe extract it to a separate function?

**Review by DifferentialOrange — COMMENTED** [review]
Thank you for your PR! Seems nice, I've found a couple of tricks that may be useful for me later. I have left a several minor comments.

File: .github/workflows/testing.yml (hunk 1 above)
Comment (oleg-jukovec) [reply .github/workflows/testing.yml:85]: Unfortunately, tt does not support installation of a specified commit. But it seems to me that the risks here are small.

File: .github/workflows/testing.yml (hunk 2 above)
Comment (oleg-jukovec) [reply .github/workflows/testing.yml:87]: In any case, we need to set up path for `cache` case.

File: request.go (hunk 4 above)
Comment (oleg-jukovec) [reply request.go:69]: Tarantool throw decoding error when I try to use `EncodeBinary` (it uses `encode_str`/`decode_str` in fact).

Comment (DifferentialOrange) [reply request.go:69]: Well, it's kinda sad, but it doesn't seems that we will be able to do anything about it

File: request.go (hunk 3 above)
Comment (oleg-jukovec) [reply request.go:47]: I have added error checking for all fields.

File: request.go (hunk 5 above)
Comment (oleg-jukovec) [reply request.go]: Thank you, updated.

File: request.go (hunk 6 above)
Comment (oleg-jukovec) [reply request.go:792]: Added.

File: request.go (hunk 7 above)
Comment (oleg-jukovec) [reply request.go:804]: Added.

File: response.go (hunk 8 above)
Comment (oleg-jukovec) [reply response.go]: Thank you, updated.

File: tarantool_test.go (hunk 9 above)
Comment (oleg-jukovec) [reply tarantool_test.go]: Great idea, updated.

File: tarantool_test.go (hunk 10 above)
Comment (oleg-jukovec) [reply tarantool_test.go]: It should be not relevant at now.

File: tarantool_test.go (hunk 11 above)
Comment (oleg-jukovec) [reply tarantool_test.go]: I split the test into several.

**Review by DifferentialOrange — APPROVED** [review]

**Review by AnaNek — APPROVED** [review]

**Pull request merged**